File: icom-bridge.py
#!/usr/bin/env python3
"""Icom IC-7610 CI-V bridge over LAN using icom-lan library.

Connects to an Icom IC-7610 via its built-in network interface and exposes
the same REST API as FlexBridge (/status, /cat/command) so server.js can
poll and forward commands without knowing the backend.

Requires: pip install icom-lan   (Python ≥ 3.11)
"""

import asyncio
import argparse
import json
import logging
import signal
import sys
import threading
import time
from http.server import HTTPServer, BaseHTTPRequestHandler
from socketserver import ThreadingMixIn
from urllib.parse import urlparse

log = logging.getLogger('icom-bridge')
logging.basicConfig(level=logging.INFO, format='%(asctime)s [%(name)s] %(message)s')

# ── Args ──────────────────────────────────────────────────────────────────────
parser = argparse.ArgumentParser(description='Icom CI-V Bridge')
parser.add_argument('--radio', default='10.0.10.112', help='Radio IP address')
parser.add_argument('--port', type=int, default=50001, help='Radio CI-V UDP port')
parser.add_argument('--username', default='', help='Network control username')
parser.add_argument('--password', default='', help='Network control password')
parser.add_argument('--civ-addr', type=lambda x: int(x, 0), default=0x98, help='CI-V address (0x98 for IC-7610)')
parser.add_argument('--status-port', type=int, default=7377, help='REST status API port')
args = parser.parse_args()

# ── State ─────────────────────────────────────────────────────────────────────
state = {
    'connected': False,
    'radio_ip': args.radio,
    'radio_model': 'IC-7610',
    'protocol_version': 'CI-V/LAN',
    'freq_hz': 14074000,
    'tx_freq_hz': 14074000,
    'rx_freq_hz': 14074000,
    'mode': 'USB',
    'split': False,
    'tx_inhibit': False,
}

radio = None  # IcomRadio instance
_loop = None  # asyncio event loop for the radio

# ── Mode mapping ──────────────────────────────────────────────────────────────
ICOM_TO_NAME = {
    'LSB': 'LSB', 'USB': 'USB', 'AM': 'AM', 'CW': 'CW', 'RTTY': 'RTTY',
    'FM': 'FM', 'CW_R': 'CW', 'RTTY_R': 'RTTY', 'PSK': 'DIGU',
    'lsb': 'LSB', 'usb': 'USB', 'am': 'AM', 'cw': 'CW', 'fm': 'FM',
}
TS2K_TO_MODE = {'1': 'LSB', '2': 'USB', '3': 'CW', '4': 'FM', '5': 'AM', '6': 'RTTY', '9': 'USB'}
MODE_TO_TS2K = {'LSB': '1', 'USB': '2', 'CW': '3', 'FM': '4', 'AM': '5', 'RTTY': '6', 'DIGU': '9', 'DIGL': '9'}


def normalise_mode(m):
    s = str(m).upper().replace('-', '_')
    return ICOM_TO_NAME.get(s, ICOM_TO_NAME.get(str(m), 'USB'))


# ── Async radio connection ────────────────────────────────────────────────────
async def radio_loop():
    global radio
    from icom_lan import create_radio, LanBackendConfig

    config = LanBackendConfig(
        host=args.radio,
        port=args.port,
        username=args.username,
        password=args.password,
        radio_addr=args.civ_addr,
    )

    while True:
        try:
            log.info(f'Connecting to {args.radio}:{args.port}...')
            radio = create_radio(config)
            await radio.connect()
            state['connected'] = True
            log.info('Connected to IC-7610!')

            # Poll loop
            while state['connected']:
                try:
                    freq = await radio.get_freq(bypass_cache=True)
                    mode_info = await radio.get_mode()
                    mode_name = normalise_mode(mode_info[0]) if mode_info else state['mode']

                    state['freq_hz'] = freq
                    state['tx_freq_hz'] = freq
                    state['rx_freq_hz'] = freq
                    state['mode'] = mode_name
                except Exception as e:
                    log.warning(f'Poll error: {e}')
                    break

                await asyncio.sleep(0.4)

        except Exception as e:
            log.warning(f'Connection error: {e}')
            state['connected'] = False
            radio = None

        state['connected'] = False
        log.info('Disconnected — reconnecting in 5s')
        await asyncio.sleep(5)


def _run_radio_loop():
    global _loop
    _loop = asyncio.new_event_loop()
    asyncio.set_event_loop(_loop)
    _loop.run_until_complete(radio_loop())


# ── CAT command handler ───────────────────────────────────────────────────────
def handle_cat(cmd):
    cmd = cmd.strip().upper().rstrip(';')
    if not cmd:
        return ''

    # FA — frequency
    if cmd.startswith('FA'):
        if len(cmd) > 2:
            try:
                hz = int(cmd[2:])
                state['freq_hz'] = hz
                state['tx_freq_hz'] = hz
                if state['connected'] and radio and _loop:
                    asyncio.run_coroutine_threadsafe(radio.set_freq(hz), _loop)
                    log.info(f'Set freq: {hz} Hz')
            except ValueError:
                pass
            return ''
        return f'FA{state["freq_hz"]:011d};'

    # MD — mode
    if cmd.startswith('MD'):
        if len(cmd) > 2:
            mode = TS2K_TO_MODE.get(cmd[2], 'USB')
            state['mode'] = mode
            if state['connected'] and radio and _loop:
                asyncio.run_coroutine_threadsafe(radio.set_mode(mode), _loop)
                log.info(f'Set mode: {mode}')
            return ''
        return f'MD{MODE_TO_TS2K.get(state["mode"], "2")};'

    # IF — transceiver info
    if cmd == 'IF':
        freq = f'{state["freq_hz"]:011d}'
        mode = MODE_TO_TS2K.get(state['mode'], '2')
        split = '1' if state['split'] else '0'
        return f'IF{freq}     +0000000000{mode}00{split}0000;'

    if cmd == 'ID': return 'ID020;'
    if cmd.startswith('AI'): return 'AI0;'

    # TX inhibit for band conflict
    if cmd == 'TX0':
        state['tx_inhibit'] = True
        log.warning('TX INHIBITED (band conflict)')
        return ''
    if cmd == 'TX1':
        state['tx_inhibit'] = False
        log.info('TX inhibit released')
        return ''

    if cmd in ('TX', 'RX'): return ''
    if cmd == 'PS': return 'PS1;'
    return '?;'


# ── REST API ──────────────────────────────────────────────────────────────────
class BridgeAPI(BaseHTTPRequestHandler):
    def log_message(self, fmt, *a):
        pass

    def _send(self, code, obj):
        body = json.dumps(obj).encode()
        self.send_response(code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(body)))
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        path = urlparse(self.path).path
        if path == '/status':
            self._send(200, {
                'connected': state['connected'],
                'radio_ip': state['radio_ip'],
                'radio_model': state['radio_model'],
                'protocol_version': state['protocol_version'],
                'freq_hz': state['freq_hz'],
                'tx_freq_hz': state['tx_freq_hz'],
                'rx_freq_hz': state['rx_freq_hz'],
                'mode': state['mode'],
                'split': state['split'],
                'tx_inhibit': state['tx_inhibit'],
            })
        else:
            self._send(404, {'error': 'not found'})

    def do_POST(self):
        path = urlparse(self.path).path
        length = int(self.headers.get('Content-Length', 0))
        body = {}
        if length:
            try:
                body = json.loads(self.rfile.read(length))
            except Exception:
                pass

        if path == '/cat/command':
            cmd = body.get('command', '').strip()
            if not cmd:
                self._send(400, {'ok': False, 'error': 'No command'})
                return
            resp = handle_cat(cmd)
            self._send(200, {'ok': True, 'response': resp})
        else:
            self._send(404, {'error': 'not found'})


class ThreadedHTTPServer(ThreadingMixIn, HTTPServer):
    daemon_threads = True
    allow_reuse_address = True


# ── Main ──────────────────────────────────────────────────────────────────────
def main():
    # Start REST API
    srv = ThreadedHTTPServer(('127.0.0.1', args.status_port), BridgeAPI)
    api_thread = threading.Thread(target=srv.serve_forever, daemon=True)
    api_thread.start()
    log.info(f'REST API on http://127.0.0.1:{args.status_port}')

    # Start radio connection in background thread
    radio_thread = threading.Thread(target=_run_radio_loop, daemon=True)
    radio_thread.start()
    log.info(f'Connecting to IC-7610 at {args.radio}:{args.port}')

    # Wait for shutdown
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        log.info('Shutting down')
        srv.shutdown()


if __name__ == '__main__':
    main()

File: icom-bridge.js
'use strict';
// ── Icom CI-V Bridge ─────────────────────────────────────────────────────────
// Connects to an Icom IC-7610 (or similar) via CI-V over TCP/IP.
// Exposes the same REST API as FlexBridge (/status, /cat/command) so
// server.js can poll and forward commands without knowing the backend.

const net  = require('net');
const http = require('http');

// ── CI-V constants ───────────────────────────────────────────────────────────
const CIV_PREAMBLE = 0xFE;
const CIV_EOM      = 0xFD;
const CIV_OK       = 0xFB;
const CIV_NG       = 0xFA;
const CTRL_ADDR    = 0xE0;

// CI-V command codes
const CMD_SET_FREQ  = 0x05;  // Set frequency (5-byte BCD)
const CMD_READ_FREQ = 0x03;  // Read frequency
const CMD_SET_MODE  = 0x06;  // Set mode + filter
const CMD_READ_MODE = 0x04;  // Read mode

// Mode byte ↔ name mapping
const ICOM_MODE_TO_NAME = {
  0x00: 'LSB', 0x01: 'USB', 0x02: 'AM', 0x03: 'CW',
  0x04: 'RTTY', 0x05: 'FM', 0x07: 'CW-R', 0x08: 'RTTY-R',
  0x17: 'DV',
};
const NAME_TO_ICOM_MODE = {};
for (const [k, v] of Object.entries(ICOM_MODE_TO_NAME)) NAME_TO_ICOM_MODE[v] = parseInt(k);

// TS-2000 mode digit ↔ CI-V mode
const TS2K_TO_ICOM = { '1': 0x00, '2': 0x01, '3': 0x03, '4': 0x05, '5': 0x02, '6': 0x04, '7': 0x07, '9': 0x01 };
const ICOM_TO_TS2K = { 0x00: '1', 0x01: '2', 0x02: '5', 0x03: '3', 0x04: '6', 0x05: '4', 0x07: '7', 0x08: '6' };

// ── BCD helpers ──────────────────────────────────────────────────────────────
function freqToIcomBCD(hz) {
  const s = String(Math.round(hz)).padStart(10, '0');
  const bytes = new Uint8Array(5);
  // 10 digits → 5 bytes, LSB first
  // s = "0014074000"  →  digits [0,0,1,4,0,7,4,0,0,0]
  // byte0 = digits[9]*16 + digits[8]  (1 Hz, 10 Hz)
  // byte1 = digits[7]*16 + digits[6]  (100 Hz, 1 kHz)
  // ...
  for (let i = 0; i < 5; i++) {
    const hi = parseInt(s[9 - i * 2 - 1]) || 0;
    const lo = parseInt(s[9 - i * 2]) || 0;
    bytes[i] = (hi << 4) | lo;
  }
  return bytes;
}

function icomBCDToFreq(bytes) {
  let hz = 0;
  for (let i = bytes.length - 1; i >= 0; i--) {
    const hi = (bytes[i] >> 4) & 0x0F;
    const lo = bytes[i] & 0x0F;
    hz = hz * 100 + hi * 10 + lo;
  }
  return hz;
}

function modeNameToFlex(name) {
  // Normalise Icom mode names to match what server.js/FlexBridge expects
  return { LSB: 'LSB', USB: 'USB', AM: 'AM', CW: 'CW', FM: 'FM',
           'CW-R': 'CW', RTTY: 'RTTY', 'RTTY-R': 'RTTY', DV: 'FM' }[name] || 'USB';
}

// ── CI-V frame builder / parser ──────────────────────────────────────────────
function buildFrame(toAddr, cmd, data) {
  const frame = [CIV_PREAMBLE, CIV_PREAMBLE, toAddr, CTRL_ADDR, cmd];
  if (data) frame.push(...data);
  frame.push(CIV_EOM);
  return Buffer.from(frame);
}

function extractFrames(buf) {
  const frames = [];
  let i = 0;
  while (i < buf.length - 1) {
    // Find preamble
    if (buf[i] !== CIV_PREAMBLE || buf[i + 1] !== CIV_PREAMBLE) { i++; continue; }
    // Find end of frame
    let end = -1;
    for (let j = i + 2; j < buf.length; j++) {
      if (buf[j] === CIV_EOM) { end = j; break; }
    }
    if (end === -1) break; // incomplete frame, keep in buffer
    // Parse: [FE FE to from cmd ...data FD]
    const to   = buf[i + 2];
    const from = buf[i + 3];
    const cmd  = buf[i + 4];
    const data = buf.slice(i + 5, end);
    frames.push({ to, from, cmd, data });
    i = end + 1;
  }
  return { frames, remaining: buf.slice(i) };
}

// ── IcomBridge class ─────────────────────────────────────────────────────────
class IcomBridge {
  constructor(opts) {
    this.radioIp    = opts.radioIp    || '10.0.10.112';
    this.radioPort  = opts.radioPort  || 50001;
    this.statusPort = opts.statusPort || 7377;
    this.civAddr    = parseInt(opts.civAddress) || 0x98;
    this.autoReconnect = opts.autoReconnect !== false;

    this.sock       = null;
    this.rxBuf      = Buffer.alloc(0);
    this.connected  = false;
    this.state      = {
      freq_hz:    14074000,
      tx_freq_hz: 14074000,
      rx_freq_hz: 14074000,
      mode:       'USB',
      mode_byte:  0x01,
      split:      false,
    };
    this.pollTimer    = null;
    this.reconnTimer  = null;
    this.httpServer   = null;
    this._log         = opts.log || console.log;
  }

  // ── Lifecycle ────────────────────────────────────────────────────────────
  start() {
    this._startHTTP();
    this.connect();
    this._log(`[Icom] Bridge starting — radio=${this.radioIp}:${this.radioPort} REST=:${this.statusPort}`);
  }

  stop() {
    clearInterval(this.pollTimer);
    clearTimeout(this.reconnTimer);
    if (this.sock) { try { this.sock.destroy(); } catch {} }
    if (this.httpServer) { try { this.httpServer.close(); } catch {} }
    this.connected = false;
    this._log('[Icom] Bridge stopped');
  }

  connect() {
    if (this.sock) { try { this.sock.destroy(); } catch {} }
    this.sock = new net.Socket();
    this.rxBuf = Buffer.alloc(0);

    this.sock.connect(this.radioPort, this.radioIp, () => {
      this.connected = true;
      this._log(`[Icom] Connected to IC-7610 at ${this.radioIp}:${this.radioPort}`);
      // Read initial state
      this._sendRead(CMD_READ_FREQ);
      this._sendRead(CMD_READ_MODE);
      // Start polling
      clearInterval(this.pollTimer);
      this.pollTimer = setInterval(() => this._poll(), 400);
    });

    this.sock.on('data', (data) => {
      this.rxBuf = Buffer.concat([this.rxBuf, data]);
      const { frames, remaining } = extractFrames(this.rxBuf);
      this.rxBuf = remaining;
      for (const f of frames) this._handleFrame(f);
    });

    this.sock.on('error', (e) => {
      this._log(`[Icom] Socket error: ${e.message}`);
    });

    this.sock.on('close', () => {
      this.connected = false;
      clearInterval(this.pollTimer);
      this._log('[Icom] Disconnected');
      if (this.autoReconnect) {
        this.reconnTimer = setTimeout(() => this.connect(), 5000);
      }
    });

    this.sock.setTimeout(5000, () => {
      if (!this.connected) {
        this._log('[Icom] Connection timeout');
        this.sock.destroy();
      }
    });
  }

  // ── CI-V I/O ─────────────────────────────────────────────────────────────
  _send(cmd, data) {
    if (!this.sock || !this.connected) return;
    const frame = buildFrame(this.civAddr, cmd, data);
    try { this.sock.write(frame); } catch {}
  }

  _sendRead(cmd) { this._send(cmd); }

  _poll() {
    this._sendRead(CMD_READ_FREQ);
    this._sendRead(CMD_READ_MODE);
  }

  _handleFrame(f) {
    // Only process frames addressed to us (or broadcast) from the radio
    if (f.from !== this.civAddr) return;
    if (f.to !== CTRL_ADDR && f.to !== 0x00) return;

    // OK / NG acknowledgement
    if (f.cmd === CIV_OK || f.cmd === CIV_NG) return;

    // Frequency data (response to read or transceive push)
    if ((f.cmd === CMD_READ_FREQ || f.cmd === CMD_SET_FREQ || f.cmd === 0x00) && f.data.length >= 5) {
      const hz = icomBCDToFreq(f.data.slice(0, 5));
      if (hz > 100000 && hz < 500000000) {
        this.state.freq_hz    = hz;
        this.state.tx_freq_hz = hz;
        this.state.rx_freq_hz = hz;
      }
      return;
    }

    // Mode data (response to read or transceive push)
    if ((f.cmd === CMD_READ_MODE || f.cmd === CMD_SET_MODE || f.cmd === 0x01) && f.data.length >= 1) {
      const modeCode = f.data[0];
      const name = ICOM_MODE_TO_NAME[modeCode];
      if (name) {
        this.state.mode      = modeNameToFlex(name);
        this.state.mode_byte = modeCode;
      }
      return;
    }
  }

  // ── Set commands ─────────────────────────────────────────────────────────
  setFreq(hz) {
    const bcd = freqToIcomBCD(hz);
    this._send(CMD_SET_FREQ, [...bcd]);
    this.state.freq_hz    = hz;
    this.state.tx_freq_hz = hz;
    this.state.rx_freq_hz = hz;
    this._log(`[Icom] Set freq: ${hz} Hz`);
  }

  setMode(modeName) {
    const code = NAME_TO_ICOM_MODE[modeName];
    if (code === undefined) {
      this._log(`[Icom] Unknown mode: ${modeName}`);
      return;
    }
    // Send mode + default filter (0x01 = FIL1)
    this._send(CMD_SET_MODE, [code, 0x01]);
    this.state.mode      = modeNameToFlex(modeName);
    this.state.mode_byte = code;
    this._log(`[Icom] Set mode: ${modeName} (0x${code.toString(16)})`);
  }

  // ── TS-2000 CAT command handler ──────────────────────────────────────────
  handleCat(cmd) {
    cmd = cmd.trim().toUpperCase().replace(/;$/, '');
    if (!cmd) return '';

    // FA — frequency
    if (cmd.startsWith('FA')) {
      if (cmd.length > 2) {
        const hz = parseInt(cmd.slice(2));
        if (!isNaN(hz) && hz > 0) this.setFreq(hz);
        return '';
      }
      return `FA${String(this.state.freq_hz).padStart(11, '0')};`;
    }

    // FB — VFO-B
    if (cmd.startsWith('FB')) {
      if (cmd.length > 2) return '';
      return `FB${String(this.state.rx_freq_hz).padStart(11, '0')};`;
    }

    // MD — mode
    if (cmd.startsWith('MD')) {
      if (cmd.length > 2) {
        const ts2k = cmd[2];
        const icomCode = TS2K_TO_ICOM[ts2k];
        if (icomCode !== undefined) {
          const name = ICOM_MODE_TO_NAME[icomCode] || 'USB';
          this.setMode(name);
        }
        return '';
      }
      const ts2k = ICOM_TO_TS2K[this.state.mode_byte] || '2';
      return `MD${ts2k};`;
    }

    // IF — transceiver info
    if (cmd === 'IF') {
      const freq  = String(this.state.freq_hz).padStart(11, '0');
      const mode  = ICOM_TO_TS2K[this.state.mode_byte] || '2';
      const split = this.state.split ? '1' : '0';
      return `IF${freq}     +0000000000${mode}00${split}0000;`;
    }

    if (cmd === 'ID') return 'ID020;';
    if (cmd.startsWith('AI')) return 'AI0;';
    if (cmd === 'TX' || cmd === 'RX') return '';
    if (cmd === 'PS') return 'PS1;';

    return '?;';
  }

  // ── REST API ─────────────────────────────────────────────────────────────
  _startHTTP() {
    this.httpServer = http.createServer((req, res) => {
      const url = new URL(req.url, `http://127.0.0.1:${this.statusPort}`);
      const path = url.pathname;

      // CORS
      res.setHeader('Access-Control-Allow-Origin', '*');
      if (req.method === 'OPTIONS') {
        res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
        res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
        res.writeHead(200); res.end(); return;
      }

      const sendJSON = (code, obj) => {
        const body = JSON.stringify(obj);
        res.writeHead(code, { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) });
        res.end(body);
      };

      if (req.method === 'GET' && path === '/status') {
        sendJSON(200, {
          connected:        this.connected,
          radio_ip:         this.radioIp,
          radio_model:      'IC-7610',
          protocol_version: 'CI-V',
          freq_hz:          this.state.freq_hz,
          tx_freq_hz:       this.state.tx_freq_hz,
          rx_freq_hz:       this.state.rx_freq_hz,
          mode:             this.state.mode,
          split:            this.state.split,
        });
        return;
      }

      if (req.method === 'POST' && path === '/cat/command') {
        let body = '';
        req.on('data', d => body += d);
        req.on('end', () => {
          try {
            const { command } = JSON.parse(body);
            if (!command) { sendJSON(400, { ok: false, error: 'No command' }); return; }
            const resp = this.handleCat(command);
            sendJSON(200, { ok: true, response: resp });
          } catch (e) {
            sendJSON(400, { ok: false, error: e.message });
          }
        });
        return;
      }

      sendJSON(404, { error: 'not found' });
    });

    this.httpServer.listen(this.statusPort, '127.0.0.1', () => {
      this._log(`[Icom] REST API on http://127.0.0.1:${this.statusPort}`);
    });
  }
}

module.exports = { IcomBridge };
